File: android/app/src/main/java/com/freebuff/mobile/PairingApi.kt
package com.freebuff.mobile

import org.json.JSONObject
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URI
import java.net.URL
import java.nio.charset.StandardCharsets

class PairingApi(rawBaseUrl: String) {
    private val baseUrl = normalizeBaseUrl(rawBaseUrl)

    fun claim(
        payload: PairingPayload,
        manualCode: String,
        deviceName: String,
        devicePublicKey: String,
    ): PairingSession {
        require(payload.baseUrl == baseUrl) { "Pairing payload endpoint changed" }
        val request = JSONObject()
            .put("pairingId", payload.pairingId)
            .put("token", payload.token)
            .put("manualCode", manualCode.trim())
            .put("deviceName", deviceName.trim())
            .put("devicePublicKey", devicePublicKey)
        return PairingSession.fromGatewayResponse(baseUrl, post("/v1/pairings/claim", request))
    }

    fun refresh(session: PairingSession): PairingSession {
        require(session.gatewayBaseUrl == baseUrl) { "Session endpoint changed" }
        val request = JSONObject()
            .put("deviceId", session.deviceId)
            .put("deviceToken", session.deviceToken)
        return PairingSession.fromGatewayResponse(
            baseUrl,
            post("/v1/sessions/refresh", request),
            deviceTokenOverride = session.deviceToken,
            deviceExpiresAtOverride = session.deviceExpiresAt,
        )
    }

    /**
     * Exchanges short-lived access token for relay-owned Secure/HttpOnly cookie.
     * Cookie is returned to native code and must be installed into WebView's
     * CookieManager; access token is never injected into page JavaScript.
     */
    fun establishWebSession(webBaseUrl: String, accessToken: String): String {
        val webOrigin = normalizeBaseUrl(webBaseUrl)
        val result = request(
            baseUrl = webOrigin,
            path = "/v1/mobile/session",
            method = "GET",
            body = null,
            headers = mapOf("Authorization" to "Bearer $accessToken"),
        )
        val cookie = result.setCookie
        require(!cookie.isNullOrBlank()) { "Relay did not return a session cookie" }
        return cookie
    }

    private fun post(path: String, body: JSONObject): JSONObject {
        return request(baseUrl, path, "POST", body, emptyMap()).json
    }

    private data class HttpResult(
        val json: JSONObject,
        val setCookie: String?,
    )

    private fun request(
        baseUrl: String,
        path: String,
        method: String,
        body: JSONObject?,
        headers: Map<String, String>,
    ): HttpResult {
        val connection = (URL("$baseUrl$path").openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 10_000
            readTimeout = 15_000
            instanceFollowRedirects = false
            doInput = true
            doOutput = body != null
            setRequestProperty("Accept", "application/json")
            setRequestProperty("Content-Type", "application/json; charset=utf-8")
            setRequestProperty("Cache-Control", "no-store")
            headers.forEach { (name, value) -> setRequestProperty(name, value) }
        }

        try {
            if (body != null) {
                val bytes = body.toString().toByteArray(StandardCharsets.UTF_8)
                connection.setFixedLengthStreamingMode(bytes.size)
                connection.outputStream.use { it.write(bytes) }
            }

            val status = connection.responseCode
            val stream = if (status in 200..299) connection.inputStream else connection.errorStream
            val responseText = stream?.bufferedReader(StandardCharsets.UTF_8)?.use { it.readText() }.orEmpty()
            val response = runCatching { JSONObject(responseText) }.getOrElse { JSONObject() }
            if (status !in 200..299) {
                throw GatewayApiException(
                    status = status,
                    message = response.optString("message", "Gateway request failed"),
                )
            }
            val setCookie = connection.headerFields.entries
                .firstOrNull { it.key?.equals("Set-Cookie", ignoreCase = true) == true }
                ?.value
                ?.firstOrNull()
            return HttpResult(response, setCookie)
        } catch (error: GatewayApiException) {
            throw error
        } catch (error: IOException) {
            throw error
        } finally {
            connection.disconnect()
        }
    }

    companion object {
        fun normalizeBaseUrl(raw: String): String {
            val uri = URI(raw.trim())
            require(uri.scheme.equals("https", ignoreCase = true)) {
                "Gateway endpoint must use HTTPS"
            }
            require(uri.userInfo == null && !uri.host.isNullOrBlank()) {
                "Gateway endpoint must not contain credentials"
            }
            return "${uri.scheme.lowercase()}://${uri.rawAuthority.lowercase()}".trimEnd('/')
        }
    }
}

class GatewayApiException(
    val status: Int,
    override val message: String,
) : IOException(message)

File: ios/FreebuffGate/PairingApi.swift
import Foundation

class PairingApi {
    let baseUrl: String

    init(rawBaseUrl: String) {
        self.baseUrl = Self.normalizeBaseUrl(rawBaseUrl)
    }

    func claim(payload: PairingPayload, deviceName: String, devicePublicKey: String) async throws -> PairingSession {
        guard payload.baseUrl == baseUrl else {
            throw PairingError.invalidUrl("Pairing payload endpoint changed")
        }
        let body: [String: Any] = [
            "pairingId": payload.pairingId,
            "token": payload.token,
            "deviceName": deviceName.trimmingCharacters(in: .whitespacesAndNewlines),
            "devicePublicKey": devicePublicKey,
        ]
        let result = try await request(
            baseUrl: baseUrl,
            path: "/v1/pairings/claim",
            method: "POST",
            body: body,
            headers: [:]
        )
        return try PairingSession.fromGatewayResponse(baseUrl: baseUrl, json: result.json)
    }

    func refresh(session: PairingSession) async throws -> PairingSession {
        guard session.gatewayBaseUrl == baseUrl else {
            throw PairingError.invalidUrl("Session endpoint changed")
        }
        let body: [String: Any] = [
            "deviceId": session.deviceId,
            "deviceToken": session.deviceToken,
        ]
        let result = try await request(
            baseUrl: baseUrl,
            path: "/v1/sessions/refresh",
            method: "POST",
            body: body,
            headers: [:]
        )
        return try PairingSession.fromGatewayResponse(
            baseUrl: baseUrl,
            json: result.json,
            deviceTokenOverride: session.deviceToken,
            deviceExpiresAtOverride: session.deviceExpiresAt
        )
    }

    /// Exchanges short-lived access token for relay-owned Secure/HttpOnly
    /// cookie. The cookie is installed into the WKWebView cookie store by
    /// native code; the access token is never injected into page JavaScript.
    /// Registers (or clears, with an empty token) the device's APNs token
    /// with the relay so it can push turn-finished notifications while the
    /// app is backgrounded.
    func uploadPushToken(session: PairingSession, token: String) async throws {
        guard session.gatewayBaseUrl == baseUrl else {
            throw PairingError.invalidUrl("Session endpoint changed")
        }
        let body: [String: Any] = ["token": token]
        _ = try await request(
            baseUrl: baseUrl,
            path: "/v1/mobile/push-token",
            method: "POST",
            body: body,
            headers: ["Authorization": "Bearer \(session.accessToken)"]
        )
    }

    func establishWebSession(webBaseUrl: String, accessToken: String) async throws -> String {
        let webOrigin = Self.normalizeBaseUrl(webBaseUrl)
        let result = try await request(
            baseUrl: webOrigin,
            path: "/v1/mobile/session",
            method: "GET",
            body: nil,
            headers: ["Authorization": "Bearer \(accessToken)"]
        )
        guard let cookie = result.setCookie, !cookie.isEmpty else {
            throw PairingError.badResponse("Relay did not return a session cookie")
        }
        return cookie
    }

    static func normalizeBaseUrl(_ raw: String) -> String {
        let text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let uri = URL(string: text), uri.scheme?.lowercased() == "https" else {
            fatalError("Gateway endpoint must use HTTPS")
        }
        guard uri.user == nil && !(uri.host ?? "").isEmpty else {
            fatalError("Gateway endpoint must not contain credentials")
        }
        var base = "\(uri.scheme!.lowercased())://\(uri.host!.lowercased())"
        if let port = uri.port {
            base += ":\(port)"
        }
        return base.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
    }

    private struct HttpResult {
        let json: [String: Any]
        let setCookie: String?
    }

    private func request(
        baseUrl: String,
        path: String,
        method: String,
        body: [String: Any]?,
        headers: [String: String]
    ) async throws -> HttpResult {
        guard let url = URL(string: "\(baseUrl)\(path)") else {
            throw PairingError.invalidUrl("Invalid gateway URL")
        }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.timeoutInterval = 15
        request.cachePolicy = .reloadIgnoringLocalAndRemoteCacheData
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("no-store", forHTTPHeaderField: "Cache-Control")
        for (name, value) in headers {
            request.setValue(value, forHTTPHeaderField: name)
        }
        if let body {
            request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        }

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw PairingError.badResponse("Gateway returned a non-HTTP response")
        }
        let status = http.statusCode
        let json: [String: Any] = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] ?? [:]
        guard (200...299).contains(status) else {
            let message = (json["message"] as? String) ?? "Gateway request failed"
            throw PairingError.http(status: status, message: message)
        }
        let setCookie = http.value(forHTTPHeaderField: "Set-Cookie")
        return HttpResult(json: json, setCookie: setCookie)
    }
}
